commit 8f285fe421bbecef85fb1f21e7b797e712c58530 (refs/changes/16/3616/1)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2018-03-10T00:07:32+00:00 (1 year, 7 months ago)
    
    TRITON-75 Allow firewall rules to be applied in order of priority
    TRITON-77 Add support for ESP and AH protocols in firewall rules
    TRITON-117 Update number of ports allowed per firewall rule

diff --git a/.gitignore b/.gitignore
index e14bb32..6798ae8 100644
--- a/.gitignore
+++ b/.gitignore
@@ -2,6 +2,8 @@
 /node_modules
 /tmp
 build
+docs/examples.md
+docs/rules.md
 docs/*.json
 docs/*.html
 docs/examples.md
diff --git a/deps/jsstyle b/deps/jsstyle
index 9600c7e..ccb145b 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit 9600c7e56c84f3a74d6e3d70c336e86f7a3e3769
+Subproject commit ccb145bab281b81e1fbebd45b5d2c084c454df5d
diff --git a/docs/index.md b/docs/index.md
index 0097bc2..33acc22 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -157,6 +157,7 @@ FROM or TO side of the rule.
 | fields     | Array of Strings | List of extra fields to return |
 | ip         | String           | Filter: IP                     |
 | owner_uuid | UUID             | Filter: Owner UUID             |
+| protocol   | String           | Filter: Protocol (e.g., "tcp") |
 | subnet     | String           | Filter: Subnet CIDR            |
 | tag        | String           | Filter: Tag                    |
 | vm         | UUID             | Filter: VM UUID                |
diff --git a/lib/rule.js b/lib/rule.js
index 974d925..5aabc8f 100644
--- a/lib/rule.js
+++ b/lib/rule.js
@@ -113,6 +113,7 @@ function ruleFromMoray(raw) {
             (raw.enabled === 'true'),
         objectclass: raw.objectclass,
         parsed: {
+            priority: raw.priority,
             action: raw.action,
             from: [],
             protocol: {
@@ -124,6 +125,10 @@ function ruleFromMoray(raw) {
         version: raw.version
     };
 
+    if (raw.protocol === 'ah' || raw.protocol === 'esp') {
+        data.parsed.protocol.targets = [ 'all' ];
+    }
+
     if (hasKey(raw, 'ports')) {
         // See "Port storage in Moray" for how ports get represented
         data.parsed.protocol.targets = raw.ports.map(function (port) {
@@ -387,6 +392,7 @@ Rule.prototype.rawMoray = function _ruleRawMoray() {
         action: this.action,
         enabled: this.enabled,
         objectclass: Rule.objectclass,
+        priority: this.priority,
         protocol: this.protocol,
         uuid: this.uuid,
         version: this.version,
diff --git a/package.json b/package.json
index 99b6239..7ea8bed 100644
--- a/package.json
+++ b/package.json
@@ -9,7 +9,7 @@
     "faucet": "0.0.1",
     "istanbul": "0.4.5",
     "eslint": "2.13.1",
-    "eslint-plugin-joyent": "1.1.0",
+    "eslint-plugin-joyent": "~1.3.0",
     "mockery": "1.4.0"
   },
   "dependencies": {
@@ -20,7 +20,7 @@
     "effluent-logger": "git+https://github.com/joshwilsdon/effluent-logger.git#d662f161a07f94045ad2afb45442931511c40e51",
     "extsprintf": "1.0.2",
     "fast-messages": "1.0.1",
-    "fwrule": "1.4.1",
+    "fwrule": "2.0.0",
     "jsprim": "1.3.1",
     "moray": "2.0.1",
     "moray-filter": "1.0.0",
diff --git a/test/integration/get.test.js b/test/integration/get.test.js
index eec4886..d3f7bef 100644
--- a/test/integration/get.test.js
+++ b/test/integration/get.test.js
@@ -44,6 +44,21 @@ var RULES = [
         enabled: true,
         global: true,
         rule: 'FROM any TO tag "foo" = "baz" ALLOW tcp PORT 5010'
+    },
+    {
+        enabled: true,
+        owner_uuid: OWNERS[1],
+        rule: 'FROM tag "a" TO tag "b" ALLOW tcp PORT 80 PRIORITY 50'
+    },
+    {
+        enabled: true,
+        owner_uuid: OWNERS[1],
+        rule: 'FROM tag "a" TO tag "b" ALLOW ah'
+    },
+    {
+        enabled: true,
+        owner_uuid: OWNERS[1],
+        rule: 'FROM tag "a" TO tag "b" ALLOW esp'
     }
 ];
 
diff --git a/test/integration/list.test.js b/test/integration/list.test.js
index fc13b1b..1abb6b4 100644
--- a/test/integration/list.test.js
+++ b/test/integration/list.test.js
@@ -71,6 +71,19 @@ var RULES = [
         owner_uuid: OWNERS[1],
         rule: 'FROM (tag "foo" = "bar" OR tag "foo" = "baz") '
             + 'TO tag "side" = "two" ALLOW tcp (PORT 5003 AND PORT 5004)'
+    },
+
+    // OWNERS[2]
+
+    {
+        enabled: true,
+        owner_uuid: OWNERS[2],
+        rule: 'FROM any TO all vms ALLOW ah'
+    },
+    {
+        enabled: true,
+        owner_uuid: OWNERS[2],
+        rule: 'FROM any TO all vms ALLOW esp'
     }
 ];
 
@@ -186,6 +199,34 @@ test('list: parsed fields', function (t) {
 });
 
 
+test('list: IPsec protocols', function (t) {
+    t.plan(2);
+
+    t.test('list: IPsec protocols - AH', function (t2) {
+        mod_rule.list(t2, {
+            params: {
+                owner_uuid: OWNERS[2],
+                protocol: 'ah'
+            },
+            exp: [
+                RULES[7]
+            ]
+        });
+    });
+
+    t.test('list: IPsec protocols - ESP', function (t2) {
+        mod_rule.list(t2, {
+            params: {
+                owner_uuid: OWNERS[2],
+                protocol: 'esp'
+            },
+            exp: [
+                RULES[8]
+            ]
+        });
+    });
+});
+
 
 // --- Teardown
 
diff --git a/test/unit/helpers.js b/test/unit/helpers.js
index 9492690..581e16d 100644
--- a/test/unit/helpers.js
+++ b/test/unit/helpers.js
@@ -33,7 +33,7 @@ var os = require('os');
 
 
 var CUR_IP = 1;
-var FWRULE_VERSION = process.env.FWRULE_VERSION || 3;
+var FWRULE_VERSION = process.env.FWRULE_VERSION || 4;
 var MOCKS_ENABLED = false;
 var MULTI_SUITE_RUN = false;
 var OWNER_UUID = mod_uuid.v4();
